components/eventing-controller/reconciler/subscription/reconciler.go

```diff
@@ -0,0 +1,964 @@
+	}
+	return svc, nil
+}
+
+func (r *Reconciler) createOrUpdateAPIRule(subscription *eventingv1alpha1.Subscription, ctx context.Context, sink url.URL, logger logr.Logger) (*apigatewayv1alpha1.APIRule, error) {
+	svcNs, svcName, err := getSvcNsAndName(sink.Host)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse svc name and ns in createOrUpdateAPIRule")
+	}
+	labels := map[string]string{
+		constants.ControllerServiceLabelKey:  svcName,
+		constants.ControllerIdentityLabelKey: constants.ControllerIdentityLabelValue,
+	}
+
+	svcPort, err := utils.GetPortNumberFromURL(sink)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to convert URL port to APIRule port")
+	}
+	var existingAPIRule *apigatewayv1alpha1.APIRule
+	existingAPIRules, err := r.getAPIRulesForASvc(ctx, labels, svcNs)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to fetch existing ApiRule for labels: %v", labels)
+	}
+	if existingAPIRules != nil {
+		existingAPIRule = r.filterAPIRulesOnPort(existingAPIRules, svcPort)
+	}
+
+	// Get all subscriptions valid for the cluster-local subscriber
+	subscriptions, err := r.getSubscriptionsForASvc(svcNs, svcName, ctx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to fetch subscriptions for the subscriber %s/%s", svcNs, svcName)
+	}
+	filteredSubscriptions := r.filterSubscriptionsOnPort(subscriptions, svcPort)
+
+	desiredAPIRule, err := r.makeAPIRule(svcNs, svcName, labels, filteredSubscriptions, svcPort)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to make an APIRule")
+	}
+
+	if existingAPIRule == nil {
+		// cleanup or update the previously used ApiRule for the subscription
+		if len(subscription.Status.APIRuleName) != 0 {
+			key := k8stypes.NamespacedName{Namespace: subscription.Namespace, Name: subscription.Status.APIRuleName}
+			previousApiRule := &apigatewayv1alpha1.APIRule{}
+
+			if err := r.Client.Get(ctx, key, previousApiRule); err != nil {
+				if !k8serrors.IsNotFound(err) {
+					return nil, err
+				}
+			} else {
+				// previous ApiRule still exists
+
+				// build a new OwnerReference list without the subscription
+				ownerReferences := make([]v1.OwnerReference, 0, len(previousApiRule.OwnerReferences))
+				for _, ownerReference := range previousApiRule.OwnerReferences {
+					if ownerReference.UID != subscription.UID {
+						ownerReferences = append(ownerReferences, ownerReference)
+					}
+				}
+
+				// delete the previous ApiRule only if the new OwnerReference list is empty
+				if len(ownerReferences) == 0 {
+					if err := r.Client.Delete(ctx, previousApiRule); err != nil {
+						return nil, err
+					}
+				} else if len(previousApiRule.OwnerReferences) > len(ownerReferences) {
+					// update the previous ApiRule only if the new OwnerReference list length is decreased
+					previousApiRule.OwnerReferences = ownerReferences
+					if err := r.Client.Update(ctx, previousApiRule); err != nil {
+						return nil, err
+					}
+				}
+			}
+		}
```

Comment:
nit: Moving this code to a function with a name similar to `handlePreviousAPIRule()` would improve readability of the code 😎

Reply:
Done.